`multiple-statements` no longer detected in a `finally:` block

### Bug description

```python
"""hi"""

try:
    pass
finally: pass  # Should complain about [multiple-statements] but doesn't
```


### Configuration

```ini
# Empty configuration
```


### Command used

```shell
/tmp/pylint/venv/bin/pylint test.py
```


### Pylint output

```
--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```


### Expected behavior

```
************* Module test
test.py:5:9: C0321: More than one statement on a single line (multiple-statements)

-------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 10.00/10, -3.33)
```

This behavior used to happen in the past.

5bdc2218ef9891e9047823c26625c62d68f2dc52 is the first bad commit

### Pylint version

```shell
pylint 3.3.0-dev0
astroid 3.3.0-dev0
Python 3.11.8 (main, May 26 2024, 00:40:16) [GCC 13.2.0]
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9759.false_negative
+++ b/doc/whatsnew/fragments/9759.false_negative
@@ -1,0 +1,3 @@
+Fix false negative for `multiple-statements` when multiple statements are present on `else` and `finally` lines of `try`.
+
+Refs #9759

--- a/pylint/checkers/format.py
+++ b/pylint/checkers/format.py
@@ -496,6 +496,10 @@ class FormatChecker(BaseTokenChecker, BaseRawFileChecker):
         prev_sibl = node.previous_sibling()
         if prev_sibl is not None:
             prev_line = prev_sibl.fromlineno
+        elif isinstance(
+            node.parent, nodes.Try
+        ) and self._is_first_node_in_else_finally_body(node, node.parent):
+            prev_line = self._infer_else_finally_line_number(node, node.parent)
         elif isinstance(node.parent, nodes.Module):
             prev_line = 0
         else:
@@ -519,6 +523,28 @@ class FormatChecker(BaseTokenChecker, BaseRawFileChecker):
                 lines.append(self._lines[line].rstrip())
             except KeyError:
                 lines.append("")
+
+    def _is_first_node_in_else_finally_body(
+        self, node: nodes.NodeNG, parent: nodes.Try
+    ) -> bool:
+        if parent.orelse and node == parent.orelse[0]:
+            return True
+        if parent.finalbody and node == parent.finalbody[0]:
+            return True
+        return False
+
+    def _infer_else_finally_line_number(
+        self, node: nodes.NodeNG, parent: nodes.Try
+    ) -> int:
+        last_line_of_prev_block = 0
+        if node in parent.finalbody and parent.orelse:
+            last_line_of_prev_block = parent.orelse[-1].tolineno
+        elif parent.handlers and parent.handlers[-1].body:
+            last_line_of_prev_block = parent.handlers[-1].body[-1].tolineno
+        elif parent.body:
+            last_line_of_prev_block = parent.body[-1].tolineno
+
+        return last_line_of_prev_block + 1 if last_line_of_prev_block else 0
 
     def _check_multi_statement_line(self, node: nodes.NodeNG, line: int) -> None:
         """Check for lines containing multiple statements."""

--- a/tests/functional/m/multiple_statements.py
+++ b/tests/functional/m/multiple_statements.py
@@ -43,3 +43,23 @@ finally:
 def concat2(arg1: str) -> str: ...
 
 def concat2(arg1: str) -> str: ...
+
+# Test for multiple statements on finally line
+try:
+    pass
+finally: pass  # [multiple-statements]
+
+# Test for multiple statements on else line
+try:
+    pass
+except:
+    pass
+else: pass  # [multiple-statements]
+
+# Test for multiple statements on else and finally lines
+try:
+    pass
+except:
+    pass
+else: pass  # [multiple-statements]
+finally: pass  # [multiple-statements]

--- a/tests/functional/m/multiple_statements.txt
+++ b/tests/functional/m/multiple_statements.txt
@@ -8,3 +8,7 @@ multiple-statements:26:30:26:59:MyException:More than one statement on a single 
 multiple-statements:27:26:27:30:MyError:More than one statement on a single line:HIGH
 multiple-statements:30:26:30:31:MyError:More than one statement on a single line:HIGH
 multiple-statements:32:26:32:31:MyError:More than one statement on a single line:HIGH
+multiple-statements:50:9:50:13::More than one statement on a single line:HIGH
+multiple-statements:57:6:57:10::More than one statement on a single line:HIGH
+multiple-statements:64:6:64:10::More than one statement on a single line:HIGH
+multiple-statements:65:9:65:13::More than one statement on a single line:HIGH

